### PR TITLE
[incubator-kie-issues#814] Fix NPE on BaseKnowledgeBuilderResultImpl.hashCode()

### LINKED ISSUE #814: NPE on BaseKnowledgeBuilderResultImpl.hashCode
NPE thrown on `BaseKnowledgeBuilderResultImpl.hashCode()`

`getMessage().hashCode()`

Root cause

AccumulateVisitor@216 instantiate an `AccumulateParsingFailedException` with no message
AccumulateVisitor@148 instantiate an `InvalidExpressionErrorResult` with the message of the previous one, that is null

`context.addCompilationError(new InvalidExpressionErrorResult(e.getMessage(), Optional.of(context.getRuleDescr())));`

`InvalidExpressionErrorResult` extends `BaseKnowledgeBuilderResultImpl`, and when `BaseKnowledgeBuilderResultImpl.hashCode()` is invoked, `getMessage()` returns `null` and `getMessage().hashCode()` throws the NPE

### PR DESCRIPTION
**Pull request by gitgabrio**
Fixes https://github.com/apache/incubator-kie-issues/issues/814

Requires https://github.com/apache/incubator-kie-kogito-runtimes/pull/3346

This PR
1. refactor `AccumulateVisitor` to avoid instantiation of `InvalidExpressionErrorResult` with `null` message
2. refacotr `BaseKnowledgeBuilderResultImpl` (and children classes) to enforce population of `message`
 
**Ports** If a forward-port or a backport is needed, paste the forward port PR here

[link](https://www.example.com)

**JIRA**: _(please edit the JIRA link if it exists)_

[link](https://www.example.com)

**referenced Pull Requests**: _(please edit the URLs of referenced pullrequests if they exist)_

* paste the link(s) from GitHub here
* link 2
* link 3 etc.

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] tests</b>

- for a <b>full downstream build</b> 
  - for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  - for <b>github actions</b> job: add the label `run_fdb`

- <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

- <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-main</b>

- for <b>quarkus lts checks</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins run quarkus-lts</b>

- for a <b>specific quarkus lts check</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-lts</b>

- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] native</b>

 - for <b>native lts checks</b>  
  Run native checks against quarkus lts branch
  Please add comment: <b>Jenkins run native-lts</b>

- for a <b>specific native lts check</b>  
  Run native checks against quarkus lts branch
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] native-lts</b>
</details>

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-7.67.x` to backport the original PR to the `7.67.x` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

### CODE REVIEW

File: drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/BaseKnowledgeBuilderResultImpl.java [hunk 1]
```diff
@@ -32,8 +32,10 @@
 public abstract class BaseKnowledgeBuilderResultImpl implements KnowledgeBuilderResult {
 
     private Resource resource;
+    private String message;
 
-    protected BaseKnowledgeBuilderResultImpl(Resource resource) {
+    protected BaseKnowledgeBuilderResultImpl(Resource resource, String message) {
+        this.message = message != null ? message : "";
```
Comment (yesamer) [inline drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/BaseKnowledgeBuilderResultImpl.java:38]: @gitgabrio ok, so passing a null message is a valid requirement, right?

Comment (gitgabrio) [reply drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/BaseKnowledgeBuilderResultImpl.java:38]: With @mariofusco we tried to be a little bit "conservative": put a stronger enforce here - e.g. with thrown exception - could have a bigger/worst impact.
As a matter of fact, all downstream builds failed, and I have to check if this modification broke something elsewhere

**Comment by mariofusco** [comment]
@gitgabrio as I feared this is also causing a couple of compilation failues in Kogito. From the log:
```
2024-01-11T14:26:21.4348889Z [ERROR] /home/runner/work/incubator-kie-drools/incubator-kie-drools/apache_incubator-kie-kogito-runtimes/jbpm/jbpm-flow-builder/src/main/java/org/jbpm/assembler/DuplicateProcess.java:[34,9] constructor ConfigurableSeverityResult in class org.drools.compiler.compiler.ConfigurableSeverityResult cannot be applied to given types;
2024-01-11T14:26:21.4353362Z [ERROR]   required: org.kie.api.io.Resource,org.kie.internal.builder.KnowledgeBuilderConfiguration,java.lang.String
2024-01-11T14:26:21.4355274Z [ERROR]   found:    org.kie.api.io.Resource,org.kie.internal.builder.KnowledgeBuilderConfiguration
2024-01-11T14:26:21.4356507Z [ERROR]   reason: actual and formal argument lists differ in length
2024-01-11T14:26:21.4360369Z [ERROR] /home/runner/work/incubator-kie-drools/incubator-kie-drools/apache_incubator-kie-kogito-runtimes/jbpm/jbpm-flow-builder/src/main/java/org/jbpm/assembler/DuplicateProcess.java:[39,19] getMessage() in org.jbpm.assembler.DuplicateProcess cannot override getMessage() in org.drools.drl.parser.BaseKnowledgeBuilderResultImpl
2024-01-11T14:26:21.4363298Z [ERROR]   overridden method is final
2024-01-11T14:26:21.4366121Z [ERROR] /home/runner/work/incubator-kie-drools/incubator-kie-drools/apache_incubator-kie-kogito-runtimes/jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/ProcessLoadError.java:[34,15] incompatible types: org.kie.api.io.Resource cannot be converted to java.lang.String
2024-01-11T14:26:21.4371716Z [ERROR] /home/runner/work/incubator-kie-drools/incubator-kie-drools/apache_incubator-kie-kogito-runtimes/jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/ProcessLoadError.java:[43,19] getMessage() in org.jbpm.compiler.ProcessLoadError cannot override getMessage() in org.drools.drl.parser.BaseKnowledgeBuilderResultImpl
2024-01-11T14:26:21.4374485Z [ERROR]   overridden method is final
```
```
2024-01-11T14:26:21.4458198Z [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project kogito-codegen-rules: Compilation failure: Compilation failure: 
2024-01-11T14:26:21.4462236Z [ERROR] /home/runner/work/incubator-kie-drools/incubator-kie-drools/apache_incubator-kie-kogito-runtimes/kogito-codegen-modules/kogito-codegen-rules/src/main/java/org/kie/kogito/codegen/rules/QueryGenerator.java:[214,49] no suitable constructor found for DroolsError(no arguments)
2024-01-11T14:26:21.4465415Z [ERROR]     constructor org.drools.drl.parser.DroolsError.DroolsError(java.lang.String) is not applicable
2024-01-11T14:26:21.4466650Z [ERROR]       (actual and formal argument lists differ in length)
2024-01-11T14:26:21.4468071Z [ERROR]     constructor org.drools.drl.parser.DroolsError.DroolsError(org.kie.api.io.Resource,java.lang.String) is not applicable
2024-01-11T14:26:21.4469605Z [ERROR]       (actual and formal argument lists differ in length)
2024-01-11T14:26:21.4473934Z [ERROR] /home/runner/work/incubator-kie-drools/incubator-kie-drools/apache_incubator-kie-kogito-runtimes/kogito-codegen-modules/kogito-codegen-rules/src/main/java/org/kie/kogito/codegen/rules/QueryGenerator.java:[219,23] getMessage() in org.kie.kogito.codegen.rules.QueryGenerator.NoBindingQuery cannot override getMessage() in org.drools.drl.parser.BaseKnowledgeBuilderResultImpl
2024-01-11T14:26:21.4477364Z [ERROR]   overridden method is final
```

File: drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/BaseKnowledgeBuilderResultImpl.java (hunk 1 above)
Comment (mariofusco) [reply drools-drl/drools-drl-parser/src/main/java/org/drools/drl/parser/BaseKnowledgeBuilderResultImpl.java:38]: @yesamer yes, a null message is allowed (as before) but now it is converted into an empty String in one centralized place.

**Comment by gitgabrio** [comment]
> @gitgabrio as I feared this is also causing a couple of compilation failues in Kogito. From the log:
> 
> ```
> 2024-01-11T14:26:21.4348889Z [ERROR] /home/runner/work/incubator-kie-drools/incubator-kie-drools/apache_incubator-kie-kogito-runtimes/jbpm/jbpm-flow-builder/src/main/java/org/jbpm/assembler/DuplicateProcess.java:[34,9] constructor ConfigurableSeverityResult in class org.drools.compiler.compiler.ConfigurableSeverityResult cannot be applied to given types;
> 2024-01-11T14:26:21.4353362Z [ERROR]   required: org.kie.api.io.Resource,org.kie.internal.builder.KnowledgeBuilderConfiguration,java.lang.String
> 2024-01-11T14:26:21.4355274Z [ERROR]   found:    org.kie.api.io.Resource,org.kie.internal.builder.KnowledgeBuilderConfiguration
> 2024-01-11T14:26:21.4356507Z [ERROR]   reason: actual and formal argument lists differ in length
> 2024-01-11T14:26:21.4360369Z [ERROR] /home/runner/work/incubator-kie-drools/incubator-kie-drools/apache_incubator-kie-kogito-runtimes/jbpm/jbpm-flow-builder/src/main/java/org/jbpm/assembler/DuplicateProcess.java:[39,19] getMessage() in org.jbpm.assembler.DuplicateProcess cannot override getMessage() in org.drools.drl.parser.BaseKnowledgeBuilderResultImpl
> 2024-01-11T14:26:21.4363298Z [ERROR]   overridden method is final
> 2024-01-11T14:26:21.4366121Z [ERROR] /home/runner/work/incubator-kie-drools/incubator-kie-drools/apache_incubator-kie-kogito-runtimes/jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/ProcessLoadError.java:[34,15] incompatible types: org.kie.api.io.Resource cannot be converted to java.lang.String
> 2024-01-11T14:26:21.4371716Z [ERROR] /home/runner/work/incubator-kie-drools/incubator-kie-drools/apache_incubator-kie-kogito-runtimes/jbpm/jbpm-flow-builder/src/main/java/org/jbpm/compiler/ProcessLoadError.java:[43,19] getMessage() in org.jbpm.compiler.ProcessLoadError cannot override getMessage() in org.drools.drl.parser.BaseKnowledgeBuilderResultImpl
> 2024-01-11T14:26:21.4374485Z [ERROR]   overridden method is final
> ```
> 
> ```
> 2024-01-11T14:26:21.4458198Z [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project kogito-codegen-rules: Compilation failure: Compilation failure: 
> 2024-01-11T14:26:21.4462236Z [ERROR] /home/runner/work/incubator-kie-drools/incubator-kie-drools/apache_incubator-kie-kogito-runtimes/kogito-codegen-modules/kogito-codegen-rules/src/main/java/org/kie/kogito/codegen/rules/QueryGenerator.java:[214,49] no suitable constructor found for DroolsError(no arguments)
> 2024-01-11T14:26:21.4465415Z [ERROR]     constructor org.drools.drl.parser.DroolsError.DroolsError(java.lang.String) is not applicable
> 2024-01-11T14:26:21.4466650Z [ERROR]       (actual and formal argument lists differ in length)
> 2024-01-11T14:26:21.4468071Z [ERROR]     constructor org.drools.drl.parser.DroolsError.DroolsError(org.kie.api.io.Resource,java.lang.String) is not applicable
> 2024-01-11T14:26:21.4469605Z [ERROR]       (actual and formal argument lists differ in length)
> 2024-01-11T14:26:21.4473934Z [ERROR] /home/runner/work/incubator-kie-drools/incubator-kie-drools/apache_incubator-kie-kogito-runtimes/kogito-codegen-modules/kogito-codegen-rules/src/main/java/org/kie/kogito/codegen/rules/QueryGenerator.java:[219,23] getMessage() in org.kie.kogito.codegen.rules.QueryGenerator.NoBindingQuery cannot override getMessage() in org.drools.drl.parser.BaseKnowledgeBuilderResultImpl
> 2024-01-11T14:26:21.4477364Z [ERROR]   overridden method is final
> ```

Yup, I'm fixing them....

**Comment by gitgabrio** [comment]
Jenkins retest this

**Comment by gitgabrio** [comment]
@mariofusco @yesamer do you know how to relaunch the build ?

**Comment by mariofusco** [comment]
> @mariofusco @yesamer do you know how to relaunch the build ?

Done.

**Review by mariofusco — APPROVED** [review]

**Review by yesamer — APPROVED** [review]

**Pull request merged**